skimage/restoration/j_invariant.py

```diff
@@ -0,0 +1,286 @@
+    stride : int, optional
+        Stride used in masking procedure that converts `denoise_function`
+        to J-invariance.
+    multichannel : bool, optional
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.
+    approximate_loss : bool, optional
+        Whether to approximate the self-supervised loss used to evaluate the
+        denoiser by only computing it on one masked version of the image.
+        If False, the runtime will be a factor of `stride**image.ndim` longer.
+
+    Returns
+    -------
+    best_denoise_function : function
+        The optimal J-invariant version of `denoise_function`.
+
+    Notes
+    -----
+
+    The calibration procedure uses a self-supervised mean-square-error loss
+    to evaluate the performance of J-invariant versions of `denoise_function`.
+    The minimizer of the self-supervised loss is also the minimizer of the
+    ground-truth loss (i.e., the true MSE error) [1]. The returned function
+    can be used on the original noisy image, or other images with similar
+    characteristics.
+
+    Increasing the stride increases the performance of `best_denoise_function`
+     at the expense of increasing its runtime. It has no effect on the runtime
+     of the calibration.
+
+    References
+    ----------
+    .. [1] J. Batson & L. Royer. Blind Denoising by Self-Supervison.
+           arxiv:1901.11365.
+
+    Examples
+    --------
+
+    >>> from skimage import color, data
+    >>> from skimage.restoration import denoise_wavelet
+    >>> import numpy as np
+    >>> img = color.rgb2gray(data.astronaut()[:50, :50])
+    >>> noisy = img + 0.5 * img.std() * np.random.randn(*img.shape)
+    >>> parameter_ranges = {'sigma': np.arange(0.1, 0.4, 0.02)}
+    >>> denoising_function = calibrate_denoiser(noisy, denoise_wavelet, parameter_ranges)
+    >>> denoised_img = denoising_function(img)
+
+    """
+    parameters_tested, losses = calibrate_denoiser_search(image,
+                                                          denoise_function,
+                                                          parameter_ranges,
+                                                          stride=stride,
+                                                          multichannel=multichannel,
+                                                          approximate_loss=approximate_loss)
+
+    idx = np.argmin(losses)
+    best_parameters = parameters_tested[idx]
+
+    best_denoise_function = lambda x: invariant_denoise(x,
+                                                        denoise_function,
+                                                        stride=stride,
+                                                        multichannel=multichannel,
+                                                        denoiser_kwargs=best_parameters)
+
+    return best_denoise_function
+
+
+def calibrate_denoiser_search(image, denoise_function, parameter_ranges, *,
```

Comment:
Should this be private? Otherwise, the docstring is not very informative about the difference with `calibrate_denoiser`

Reply:
I had hard time to figure out why we have two functions with close behaviors. If the purpose is to get information about the process, then I think this must be an option to set, not a different function to use.

Reply:
Thanks for raising this. I think the docstring needs updating, and maybe the title, but it's still a useful public function. One function, `calibrate_denoiser`, returns a denoising function, ready to apply to new images, while `calibrate_denoiser_search` returns parameter combinations and corresponding losses.

Probably what we need here is for this API to make it easy to swap out search algorithms in the future, so that we don't have to rely on grid search forever, and if we do add other search strategies, we don't have to have a deprecation cycle...